File: solutions/haskell/protein-translation/src/ProteinTranslation.hs
module ProteinTranslation(proteins) where

protein :: String -> Maybe String
protein "AUG" = Just "Methionine"
protein "UUC" = Just "Phenylalanine"
protein "UUU" = Just "Phenylalanine"
protein "UUG" = Just "Leucine"
protein "UUA" = Just "Leucine"
protein "UCG" = Just "Serine"
protein "UCA" = Just "Serine"
protein "UCC" = Just "Serine"
protein "UCU" = Just "Serine"
protein "UAC" = Just "Tyrosine"
protein "UAU" = Just "Tyrosine"
protein "UGC" = Just "Cysteine"
protein "UGU" = Just "Cysteine"
protein "UGG" = Just "Tryptophan"
protein "UAG" = Just "Stop"
protein "UAA" = Just "Stop"
protein "UGA" = Just "Stop"
protein _ = Nothing

proteins :: String -> Maybe [String]
proteins "" = Just []
proteins s = case protein first of
  Nothing -> Nothing
  Just "Stop" -> Just []
  Just amino -> case proteins remaining of
    Nothing -> Nothing
    Just rest -> Just $ amino : rest
  where
    (first, remaining) = splitAt 3 s

File: solutions/ruby/protein-translation/protein_translation.rb
class InvalidCodonError < StandardError
end

module Translation
  CODONS = {
    'AUG' => 'Methionine',
    'UUU' => 'Phenylalanine',
    'UUC' => 'Phenylalanine',
    'UUA' => 'Leucine',
    'UUG' => 'Leucine',
    'UCU' => 'Serine',
    'UCC' => 'Serine',
    'UCA' => 'Serine',
    'UCG' => 'Serine',
    'UAU' => 'Tyrosine',
    'UAC' => 'Tyrosine',
    'UGU' => 'Cysteine',
    'UGC' => 'Cysteine',
    'UGG' => 'Tryptophan',
    'UAA' => :STOP,
    'UAG' => :STOP,
    'UGA' => :STOP
  }.freeze
  private_constant :CODONS

  def self.of_rna(strand)
    strand
      .chars
      .each_slice(3)
      .map(&:join)
      .take_while { CODONS.fetch(_1, :INVALID) != :STOP }
      .map do |codon|
        raise InvalidCodonError unless CODONS.include?(codon)

        CODONS[codon]
      end
  end
end
